### PR TITLE
fix set access

### PR DESCRIPTION
**Pull request by haseebeqx**
fix for #8697 do not allow to set access to non existing websites
this is a minor security issue

### CODE REVIEW

File: plugins/UsersManager/Model.php [hunk 1]
```diff
@@ -245,11 +245,14 @@ public function userEmailExists($userEmail)
...
-                      "access" => $access)
-            );
+            $count = $this->getDb()->fetchOne("SELECT count(*) FROM " . Common::prefixTable("site"). " WHERE idsite = ?",$idsite);
+            if($count != 0){
+                 $this->getDb()->insert(Common::prefixTable("access"),
+                    array("idsite" => $idsite,
+                         "login"  => $userLogin,
+                         "access" => $access)
+                );
+            }
```
Comment (tsteur) [inline plugins/UsersManager/Model.php]: Our models are supposed to be very stupid, not having any logic. It's not really like a model more like a DAO. Also the users manager should not know about the sites manager or the data structure of sites (in the case it should not having to know that there is a `site` table etc).

Ideally the logic would be implemented in the `UsersManager\API` method `setUserAccess()`. There you could do something like 

```
foreach ($idSites as $idsite) {
    $site = Piwik\API\Request::processRequest('SitesManager.getSiteFromId', array('idSite' => $idSite));
    if (!empty($site)) {
         $this->model->addUserAccess($userLogin, $access, $idSite);
   }
}
```

I have not tested the code but something like that should work. I know it's a bit annoying as one would have to do it also around `deleteUserAccess()` etc. Maybe the code could be simplified eg `$idSites = $this->getOnlyActuallyExistingIdSites($idSites)` at the beginning of `setUserAccess` API method.

Comment (haseebeqx) [inline plugins/UsersManager/Model.php]: i don't understand what you are saying

Comment (tsteur) [inline plugins/UsersManager/Model.php]: What I meant is there should be no `"SELECT count(*) FROM " . Common::prefixTable("site"). " WHERE idsite = ?",$idsite` in this method. This method that you changed should stay unchanged. 

Instead, wherever `addUserAccess()` is called, one should ask the SitesManager API (`Piwik\API\Request::processRequest('SitesManager.getSiteFromId', array('idSite' => $idSite));`) whether the site actually exists. For example:

```php
foreach ($idSites as $idsite) {
    $site = Piwik\API\Request::processRequest('SitesManager.getSiteFromId', array('idSite' => $idSite));
    if (!empty($site)) {
         $this->model->addUserAccess($userLogin, $access, $idSite);
   }
}
```

Comment (tsteur) [inline plugins/UsersManager/Model.php]: Hey @haseebeqx are you still working on the PR? Can I maybe help you somehow or explain it in a different way?

Comment (haseebeqx) [inline plugins/UsersManager/Model.php]: I am not.

Comment (tsteur) [inline plugins/UsersManager/Model.php]: Thank you for letting us know. I will close the PR for now. Cheers for starting it, we will try to apply the other mentioned changes some time and fix it.

**Comment by tsteur** [comment]
PR is incomplete and user is not working on it anymore. This PR is a good start and we only need to apply a few changes to fix this issue.

**Pull request closed without merging**